### DISCUSSION TITLE
high-complexity equations fitting

### DISCUSSION THREAD

**Discussion by Enqiliu125** [MilesCranmer/PySR#634] [Q&A]
Dear Everynoe,

I am currently using PySR to fit high-complexity equations such as the Arrhenius equation like y=A*x^n*exp(-E/R/x). However, I am encountering a problem where the loss function is extremely high, resulting in a failure to fit the model properly. Could you please suggest any methods to avoid such issues? Additionally, can you provide guidance on how to implement constraints or limitations within the fitting process?

Below is my code:
```ruby
import pysr
import sympy
import numpy as np
from matplotlib import pyplot as plt
from pysr import PySRRegressor
from sklearn.model_selection import train_test_split
from math import exp
A = 3.55*10**15
n = -0.41
E = 16.6  #  J/mol
R = 8.314  #  J/(mol*K)
X = np.linspace(250, 1250, 1000)
y = A*X**n*np.exp(-E/R/X)
default_pysr_params = dict(
    populations=30,
    model_selection="best",
    #random_state=42,
    #proce=0,
    #deterministic=True
)
# Learn equations
model = PySRRegressor(
    niterations=300,
    binary_operators=["+", "-", "*", "/","^"],
    unary_operators=["exp"],
    #extra_sympy_mappings={"inv": lambda x:np.exp(1/x) },
    **default_pysr_params,
)
X_train, X_test, y_train, y_test = train_test_split(X.reshape(-1, 1), y, test_size=0.2, random_state=42)
model.fit(X_train, y_train)
model.sympy()
```

**Comment by MilesCranmer** [comment]
Hi @Enqiliu125, did you try the tips on https://astroautomata.com/PySR/tuning/?

**Comment by pukpr** [comment]
For Arrhenius-type equations, perhaps a good idea to log-scale values first to dynamically equalize values across the range?

**Reply by pukpr** [reply]
> "For Arrhenius-type equations, perhaps a good idea to log-scale values first to dynamically equalize values across the range?"

I wonder if the user ever saw the response above.  This should be a good idea, correct?  It's why they invented semi-log graph paper and a log option on plots.

**Reply by pukpr** [reply]
I bring this up again because this class of (x,y) data, y =  c×exp(g(x)), has the trivial  log of both sides helper function, log(y) = g(x)+log(c). This equalizes the dynamic range allowing a balanced fit for better reverse engineering what g(x) is.  However, for something like y = c×sin(g(x)), the trick of applying arcsin is challenging. asin(y/c) = g(x) is multi-valued and can vary wildly depending on the value of c.   SSo, there should be some "dual capability" of optimizing a solution for  both LHS and RHS of asin(y/c) = g(x).    Or in general,  anything like y = c×f(g(x)), where f and g are dual optimization paths.
